pyro/infer/renyi_elbo.py

```diff
@@ -0,0 +1,244 @@
+            elbo_particle = 0
+            surrogate_elbo_particle = 0
+
+            # compute elbo and surrogate elbo
+            for name, site in model_trace.nodes.items():
+                if site["type"] == "sample":
+                    if is_vectorized:
+                        log_prob_sum = site["log_prob"].reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = site["log_prob_sum"]
+                    elbo_particle = elbo_particle + log_prob_sum.detach()
+                    surrogate_elbo_particle = surrogate_elbo_particle + log_prob_sum
+
+            for name, site in guide_trace.nodes.items():
+                if site["type"] == "sample":
+                    log_prob, score_function_term, entropy_term = site["score_parts"]
+                    if is_vectorized:
+                        log_prob_sum = log_prob.reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = site["log_prob_sum"]
+
+                    elbo_particle = elbo_particle - log_prob_sum.detach()
+
+                    if not is_identically_zero(entropy_term):
+                        surrogate_elbo_particle = surrogate_elbo_particle - log_prob_sum
+
+                    if not is_identically_zero(score_function_term):
+                        surrogate_elbo_particle = (surrogate_elbo_particle +
+                                                   (self.alpha / (1. - self.alpha)) * log_prob_sum)
+
+            if is_identically_zero(elbo_particle):
+                if tensor_holder is not None:
+                    elbo_particle = tensor_holder.new_zeros(tensor_holder.shape)
+                    surrogate_elbo_particle = tensor_holder.new_zeros(tensor_holder.shape)
+            else:  # elbo_particle is not None
+                if tensor_holder is None:
+                    tensor_holder = elbo_particle.new_empty(elbo_particle.shape)
+                    # change types of previous `elbo_particle`s
+                    for i in range(len(elbo_trace)):
+                        elbo_trace[i] = tensor_holder.new_zeros(tensor_holder.shape)
+                        surrogate_elbo_trace[i] = tensor_holder.new_zeros(tensor_holder.shape)
+
+            elbo_trace.append(elbo_particle)
+            surrogate_elbo_trace.append(surrogate_elbo_particle)
+
+        if tensor_holder is None:
+            return 0.
+
+        if self.num_particles == 1:
+            elbo = elbo_trace[0]
+            surrogate_elbo_particles = surrogate_elbo_trace[0]
+        else:
+            if is_vectorized:
+                elbo_particles = elbo_trace[0]
+                surrogate_elbo_particles = surrogate_elbo_trace[0]
+            else:
+                elbo_particles = torch.stack(elbo_trace)
+                surrogate_elbo_particles = torch.stack(surrogate_elbo_trace)
+
+            elbo_particles_scaled = (1. - self.alpha) * elbo_particles
```

Comment:
Related to above, can we simplify this to only compute weights at the very end:
```
   scaled_weights = (1 - self.alpha) * elbo_particles
   weights = (scaled_weights - log_sum_exp(scaled_weights) + math.log(self.num_particles)).exp()
```

Reply:
This computation is for `elbo` also. I don't think your suggestion will simplify things. Please correct me if I am wrong.

Reply:
Once we have the weights, they can be used to return the elbo loss,  say something like this? Not sure about how much simpler it is, but seems more intuitive. :)

```
        scaled_weights = (1 - self.alpha) * elbo_particles
        log_mean_weight = log_sum_exp(scaled_weights) - math.log(self.num_particles)
        weights = (scaled_weights - log_mean_weight).exp()

        if trainable_params and getattr(surrogate_particles, 'requires_grad', False):
            surrogate_elbo = -(weights * surrogate_particles).sum() / self.num_particles
            surrogate_elbo.backward()

        loss = - (1. / (1. - self.alpha) * log_mean_weight)

        warn_if_nan(loss, "loss")
        return loss
```